pkg/gce-cloud-provider/compute/gce-compute.go

```diff
@@ -19,128 +19,365 @@ import (
 	"sigs.k8s.io/gcp-compute-persistent-disk-csi-driver/pkg/common"
 )
 
+const (
+	operationStatusDone = "DONE"
+)
+
 type GCECompute interface {
-	// Getters
-	GetProject() string
-	GetZone() string
 	// Disk Methods
-	GetDiskOrError(ctx context.Context, volumeZone, volumeName string) (*compute.Disk, error)
-	GetAndValidateExistingDisk(ctx context.Context, configuredZone, name, diskType string, reqBytes, limBytes int64) (exists bool, err error)
-	InsertDisk(ctx context.Context, zone string, diskToCreate *compute.Disk) (*compute.Operation, error)
-	DeleteDisk(ctx context.Context, zone, name string) (*compute.Operation, error)
-	AttachDisk(ctx context.Context, zone, instanceName string, attachedDisk *compute.AttachedDisk) (*compute.Operation, error)
-	DetachDisk(ctx context.Context, volumeZone, instanceName, volumeName string) (*compute.Operation, error)
-	GetDiskSourceURI(disk *compute.Disk, zone string) string
-	GetDiskTypeURI(zone, diskType string) string
-	WaitForAttach(ctx context.Context, zone, diskName, instanceName string) error
+	GetDisk(ctx context.Context, volumeKey *meta.Key) (*CloudDisk, error)
+	ValidateExistingDisk(ctx context.Context, disk *CloudDisk, diskType string, reqBytes, limBytes int64) error
+	InsertDisk(ctx context.Context, volKey *meta.Key, diskType string, capBytes int64, capacityRange *csi.CapacityRange, replicaZones []string) error
+	DeleteDisk(ctx context.Context, volumeKey *meta.Key) error
+	AttachDisk(ctx context.Context, disk *CloudDisk, volKey *meta.Key, readWrite, diskType, instanceZone, instanceName string) error
+	DetachDisk(ctx context.Context, volKey *meta.Key, instanceZone, instanceName string) error
+	GetDiskSourceURI(disk *CloudDisk, volKey *meta.Key) string
+	GetDiskTypeURI(volKey *meta.Key, diskType string) string
+	WaitForAttach(ctx context.Context, volKey *meta.Key, instanceZone, instanceName string) error
+	// Regional Disk Methods
+	GetReplicaZoneURI(zone string) string
 	// Instance Methods
 	GetInstanceOrError(ctx context.Context, instanceZone, instanceName string) (*compute.Instance, error)
-	// Operation Methods
-	WaitForOp(ctx context.Context, op *compute.Operation, zone string) error
+	// Zone Methods
+	ListZones(ctx context.Context, region string) ([]string, error)
 }
 
-func (cloud *CloudProvider) GetProject() string {
-	return cloud.project
+func (cloud *CloudProvider) ListZones(ctx context.Context, region string) ([]string, error) {
+	zones := []string{}
+	zoneList, err := cloud.service.Zones.List(cloud.project).Filter(fmt.Sprintf("region eq .*%s$", region)).Do()
+	if err != nil {
+		return nil, fmt.Errorf("failed to list zones in region %s: %v", region, err)
+	}
+	for _, zone := range zoneList.Items {
+		zones = append(zones, zone.Name)
+	}
+	return zones, nil
+
 }
 
-func (cloud *CloudProvider) GetZone() string {
-	return cloud.zone
+func (cloud *CloudProvider) GetDisk(ctx context.Context, key *meta.Key) (*CloudDisk, error) {
+	switch key.Type() {
+	case meta.Zonal:
```

Comment:
Instead of having case statements for the two types of disks, I'm wondering if it would be simpler to define a generic Disk interface, and have a zonal implementation, and a regional implementation.

Reply:
I think there are merits to both ways, although the current method I still like better for 3 reasons:
1) This isn't just an interface for disks, there are some other "types" of calls as well that wouldn't cleanly fit in a disk interface and I would rather just maintain 1 interface for "cloud provider"
2) Some calls are not different between regional/zonal, I believe in go since there is no inheritance we would just have to copy/paste that code if there were two types of disks
3) It's already like this and would take some effort to change. Without a strong reason otherwise it would be better to spend that effort elsewhere.